server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,314 @@
+        team_b.extend(search for search in searches if search not in team_a)
+
+        combined_team_a = CombinedSearch(*team_a)
+        combined_team_b = CombinedSearch(*team_b)
+        self._logger.debug("made teams: Average rating: %s target strength: %s", avg, team_target_strength)
+        self._logger.debug("team a: %s cumulated rating: %s average rating: %s",
+                           team_a, combined_team_a.cumulated_rating, combined_team_a.average_rating)
+        self._logger.debug("team b: %s cumulated rating: %s average rating: %s",
+                           team_b, combined_team_b.cumulated_rating, combined_team_b.average_rating)
+        if not len(combined_team_a.players) == team_size:
+            raise UnevenTeamsException()
+        if not len(combined_team_b.players) == team_size:
+            raise UnevenTeamsException()
+        return combined_team_a, combined_team_b
+
+    def _run_karmarkar_karp_algorithm(self, searches):
+        self._logger.debug("Running Karmarkar-Karp to partition the teams")
+        # Further reading: https://en.wikipedia.org/wiki/Largest_differencing_method
+        # Karmarkar-Karp works only for positive integers. By adding 5000 to the rating of each player
+        # we also strongly incentivise the algorithm to give both teams the same number of players
+        containers = SortedList([Container(5000 * len(s.players) + s.cumulated_rating, [s]) for s in searches],
+                                key=lambda c: c.rating)
+
+        elem1 = containers.pop()
+        elem2 = containers.pop()
+        while True:
+            #  elem1 is always bigger than elem2
+            container = Container(elem1.rating - elem2.rating, [elem1, elem2])
+            containers.add(container)
+            elem1 = containers.pop()
+            try:
+                elem2 = containers.pop()
+            except IndexError:
+                break
+        self._logger.debug("Rating disparity: %s", elem1.rating)
+
+        # We now need to open all containers again to get to the searches. A container can hold a single
+        # search or two other containers, so we differentiate the two cases by the length of the content array.
+        # Because each container represent the difference of the two containing elements they have to go in
+        # different teams (The higher one into the team the container is in).
+        team_a = []
+        team_b = []
+        containers_a = []
+        containers_b = []
+        containers_a.append(elem1.content[0])
+        containers_b.append(elem1.content[1])
+        while containers_a or containers_b:
+            for e in containers_a:
+                if len(e.content) == 2:
+                    containers_a.append(e.content[0])
+                    containers_b.append(e.content[1])
+                else:
+                    team_a.append(e.content[0])
+                containers_a.remove(e)
```

Comment:
Is it really safe to remove from and append to `containers_a` while iterating over it? A loop that pops from `containers_a` might be safer.